observer.py

```diff
@@ -0,0 +1,311 @@
+from __future__ import annotations
+import operator
+import os
+import json
+import time
+import logging
+import logging.config
+from typing import Optional
+import asyncio
+
+from discord_webhook import DiscordEmbed, DiscordWebhook
+
+from data_classes import EmbedData, TokenSale
+from api_calls import ApiCalls
+
+
+with open("configs/log_config.json", "r", encoding="UTF-8") as stream:
+    config = json.load(stream)
+logging.config.dictConfig(config)
+logger = logging.getLogger("standard")
+with open("configs/config.json", encoding="UTF-8") as g:
+    configs = json.load(g)
+
+
+class SaleFinderSubject:
+    """
+    Manages observers and filter. Gets sales data, processes it and notifies observers.
+    """
+
+    token_sale_list: list[TokenSale]
+    raw_sales_data: dict = {}
+    avax_price_in_usd: float = 0
+    sale_to_notify: TokenSale
+    embed_data: EmbedData
+    _observers: list[FilteredDiscordObserver] = []
+    _observed_collections: list[str] = []
+    _last_notified_transactions: list[str]
+
+    def __init__(self) -> None:
+        with open("configs/lastNotifiedTransactions.json", encoding="UTF-8") as file:
+            self._last_notified_transactions = json.load(file)
+        self.JOEPEGS_API_KEY: Optional[str] = os.getenv("JOEPEGS_API_KEY")
+
+    def attach(self, observer: FilteredDiscordObserver) -> None:
+        self._observers.append(observer)
+        logger.debug("Subject: Attached an observer.")
+
+    def detach(self, observer: FilteredDiscordObserver) -> None:
+        self._observers.remove(observer)
+        logger.debug("Subject: Detached an observer.")
+
+    def set_filter(self, collection_ids: list[str]) -> None:
+        self._observed_collections = [
+            collection.lower() for collection in collection_ids
+        ]
+        logger.warning(
+            "Filter was set up. Will only notify about these collections: %s",
+            self._observed_collections,
+        )
+
+    async def run(self) -> None:
+        """
+        Get sales data, process it, notify observers and save notified.
+        """
+        logger.debug("Run method executing")
+        self.api_calls = ApiCalls()

```

Comment:
`ApiCalls` doesn't need to be assigned to a class variable here.

Reply:
i use close_client() method in `main.py` finally clause to safely close connection every time to avoid leaking + in get_single_sale_data_from method in `observer.py`, so I need it as class variable really